docs/contribute-add-feature.md

```diff
@@ -0,0 +1,460 @@
+
+:::{tip}
+You can hover your mouse cursor over the link to [`packages/myst-roles/src/abbreviation.ts`](https://github.com/executablebooks/mystmd/blob/main/packages/myst-roles/src/abbreviation.ts) to see the contents of the file.
+:::
+
+We can see that `abbrevationRole` is annotated with the type `RoleSpec`. This is the basic type of a role declaration defined by the MyST specification. There are a number of important fields, such as the `name`, `alias`, and `body`. Our role will have the name `word-count`, and knowing that, we can define a barebones implementation that doesn't do anything! Let's add a new source file[^src] `word-count.ts` in the `myst-roles` package, and write the following:
+```{code-block} typescript
+:filename: packages/myst-roles/src/word-count.ts
+
+import type { RoleSpec, RoleData, GenericNode } from 'myst-common';
+
+export const wordCountRole: RoleSpec = {
+  name: 'word-count',
+  body: {
+    type: String,
+    required: true,
+  },
+  run(data: RoleData): GenericNode[] {
+    // TODO!
+    return [];
+  }
+};
+```
+
+In order to determine what should our `run` function should do, we must understand how MyST documents are built. MyST generates a website or article from a MyST Markdown file in roughly three phases, shown in the diagram below.
+:::{mermaid}
+
+graph LR;
+parse[Parse Markdown into AST]
+transform[Transform AST into New ASTs]
+export[Export AST to Website/Article]
+
+parse --> transform --> export
+:::
+At the heart of MyST is the AST, defined by the [MyST Specification](https://spec.mystmd.org/), which serves as a structured representation of a document. Whilst directives, roles, and fragments of Markdown syntax are individually processed to build this AST, transforms are performed on the _entire_ tree, i.e. over the entire document. As computing the word-count requires access to the entire document, it is clear that all of the logic of our new feature will need to be implemented as a transform. Therefore, our role definition will be very simple - generating a simple AST node that our yet-unwritten transform can later replace:
+
+```{code-block} typescript
+:linenos:
+:filename: packages/myst-roles/src/word-count.ts
+:emphasize-lines: 10,11,12,13,14,15
+
+import type { RoleSpec, RoleData, GenericNode } from 'myst-common';
+
+export const wordCountRole: RoleSpec = {
+  name: 'word-count',
+  body: {
+    type: String,
+    required: true,
+  },
+  run(data: RoleData): GenericNode[] {
+    return [
+        {
+          type: 'wordCount',
+          value: data.body as string
+        }
+    ];
+  }
+};
+```
+
+Our new role is not-yet ready to be used. We next need to tell MyST that it should be included in the main program. To do this, first we must import the role in `packages/myst-roles/src/index.ts`
```

Comment:
````suggestion
```

Up to this point, the role simply copies the contents of the body to an `AST` node. The rest of the implementation continues in ...

However, our new role is not-yet included in the main MyST program. To do this, first we must import the role in `packages/myst-roles/src/index.ts`
````

A short summary of what is implemented up to there and what is implemented in the next section.

Reply:
I've edited this in favour of adding more explanation to the role